### PR TITLE
[Controller] Add controller interface and implementations

### PR DESCRIPTION
**Pull request by tryuan99**
This PR adds the implementation for the controllers, including `PnController`, `ApnController`, and `WaypointController`.  Previously, the `WaypointController` would set the acceleration, so that the agent would achieve the desired velocity vector within one simulation step, which I think is brittle and violates abstraction barriers.  Instead, the `WaypointController` now simply uses the maximum forward and normal accelerations to try to align its velocity vector to point towards the target.  Unit tests are provided for all controllers.

### CODE REVIEW

File: Assets/Tests/EditMode/WaypointControllerTests.cs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    Assert.AreEqual(10, _controller.Plan().z);
+  }
+
+  [Test]
+  public void Plan_UsesMaximumNormalAcceleration() {
+    _agent.HierarchicalAgent.TargetModel =
+        new FixedHierarchical(position: new Vector3(0, 1, 1), velocity: new Vector3(0, 1, -1),
+                              acceleration: new Vector3(0, 1, 0));
+    Assert.AreEqual(5, _controller.Plan().y);
+  }
```
[the remark this answers is not in the document]
Comment (tryuan99) [reply Assets/Tests/EditMode/WaypointControllerTests.cs]: No.

File: Assets/Scripts/Controller/WaypointController.cs [hunk 2]
```diff
@@ -1,22 +1,27 @@
...
-    _desiredSpeed = desiredSpeed;
-  }
+  public WaypointController(IAgent agent) : base(agent) {}
+
+  // Controller-dependent implementation of the control law.
+  protected override Vector3 Plan(in Transformation relativeTransformation) {
+    Vector3 relativePosition = relativeTransformation.Position.Cartesian;
+    if (relativeTransformation.Position.Range < _accelerationCutoffDistanceSqr) {
+      return Vector3.zero;
+    }
```
Comment (daniellovell) [inline Assets/Scripts/Controller/WaypointController.cs]: This distance check is suspicious to me - `_accelerationCutoffDistanceSqr` the name suggest a squared distance but it's being compared against Range which is a linear distance?

File: Assets/Scripts/Controller/WaypointController.cs [hunk 3]
```diff
@@ -1,22 +1,27 @@
...
+    Vector3 forward = Agent.transform.forward;
+    Vector3 forwardAccelerationInput = forward * Agent.MaxForwardAcceleration();
 
-    // Calculate the acceleration needed to reach the desired velocity within one time step.
-    Vector3 accelerationInput =
-        (desiredVelocity - _agent.GetVelocity()) / (float)Time.fixedDeltaTime;
-    return accelerationInput;
+    // Steer as hard as possible towards the waypoint.
+    Vector3 normalAcceleration = Vector3.ProjectOnPlane(relativePosition, forward).normalized;
+    Vector3 normalAccelerationInput = normalAcceleration * Agent.MaxNormalAcceleration();
```
Comment (daniellovell) [inline Assets/Scripts/Controller/WaypointController.cs]: nit: NRE
The `WaypointController` calls `Agent.MaxForwardAcceleration()` and `Agent.MaxNormalAcceleration()` without ensuring `Agent.StaticConfig` is not null. While `AgentBase.MaxForwardAcceleration()` has null-coalescing (`??`), the test setup in WaypointControllerTests.cs manually sets `StaticConfig` - this is a dependency that's not obvious.

File: Assets/Tests/EditMode/ApnControllerTests.cs [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  }
+
+  [Test]
+  public void Plan_TargetAtBoresight_NonzeroClosingVelocity_MovingLeft() {
+    _agent.HierarchicalAgent.TargetModel =
+        new FixedHierarchical(position: new Vector3(0, 0, 1), velocity: new Vector3(-1, 0, -1),
+                              acceleration: new Vector3(0, 1, 0));
+    Assert.AreEqual(new Vector3(-1, 0.5f, 0), _controller.Plan());
+  }
+}
```
Comment (daniellovell) [inline Assets/Tests/EditMode/ApnControllerTests.cs]: nit: magic numbers in tests

```suggestion
  [SetUp]
  public void SetUp() {
    _agent = new GameObject("Agent").AddComponent<AgentBase>();
    Rigidbody agentRb = _agent.gameObject.AddComponent<Rigidbody>();
    _agent.transform.rotation = Quaternion.LookRotation(new Vector3(0, 0, 1));
    InvokePrivateMethod(_agent, "Awake");
    _agent.HierarchicalAgent = new HierarchicalAgent(_agent);
    _controller = new ApnController(_agent, gain: 1);
  }

  [Test]
  // PN component: gain(1) * closingVel(1) * losRateEl(1) = 1 (up)
  // APN feedforward: gain(1)/2 * targetAccel.y(1) = 0.5 (up)
  // Total: [0, 1.5, 0]
  Assert.AreEqual(new Vector3(0, 1.5f, 0), _controller.Plan());
  public void Plan_TargetAtBoresight_NonzeroClosingVelocity_MovingUpwards() {
    _agent.HierarchicalAgent.TargetModel =
        new FixedHierarchical(position: new Vector3(0, 0, 1), velocity: new Vector3(0, 1, -1),
                              acceleration: new Vector3(0, 1, 0));
    Assert.AreEqual(new Vector3(0, 1.5f, 0), _controller.Plan());
  }

  [Test]
  public void Plan_TargetAtBoresight_NonzeroClosingVelocity_MovingLeft() {
    _agent.HierarchicalAgent.TargetModel =
        new FixedHierarchical(position: new Vector3(0, 0, 1), velocity: new Vector3(-1, 0, -1),
                              acceleration: new Vector3(0, 1, 0));
    Assert.AreEqual(new Vector3(-1, 0.5f, 0), _controller.Plan());
  }
}
```

File: Assets/Tests/EditMode/PnControllerTests.cs
Comment (daniellovell) [inline Assets/Tests/EditMode/PnControllerTests.cs]: Some more suggested tests for edge cases. Could ask your favorite AI coding agent to generate them for you:
- Target moving perpendicular at exactly 90deg (tests spiral detection)
- Target with negative closing velocity (receding)
- Target at extreme elevations (near +/-90deg)
- Multiple sequential calls to verify state consistency

File: Assets/Tests/EditMode/ApnControllerTests.cs
Comment (daniellovell) [inline Assets/Tests/EditMode/ApnControllerTests.cs]: Some more suggested tests for edge cases:
- Same as PnController, plus:
- Target with zero acceleration
- Target with large acceleration relative to velocity
- Verify APN adds exactly gain/2 * normalAcceleration to PN output

File: Assets/Scripts/Controller/IController.cs [hunk 5]
```diff
@@ -5,9 +5,11 @@
 // The controller determines the acceleration input given the agent's current state and its intended
 // target or waypoint.
 public interface IController {
+  IAgent Agent { get; set; }
```
Comment (daniellovell) [inline Assets/Scripts/Controller/IController.cs]: Recommend either 1) init-only or 2) constructor-only assignment and remove the setter
```suggestion
  IAgent Agent { get; init; }
```

File: Assets/Scripts/Controller/WaypointController.cs (hunk 2 above)
Comment (tryuan99) [reply Assets/Scripts/Controller/WaypointController.cs]: Fixed, thanks!

File: Assets/Scripts/Controller/PnController.cs [hunk 6]
```diff
@@ -1,32 +1,29 @@
...
+// sight's rotation rate to steer the agent towards its target.
+public class PnController : ControllerBase {
   // Negative closing velocity turn factor.
-  protected const float negativeClosingVelocityTurnFactor = 100f;
+  private const float _negativeClosingVelocityTurnFactor = 100f;
 
-  // Minimum line-of-sight rate.
-  protected const float minimumLosRate = 0.2f;
+  // Angular threshold in degrees for detecting spiral behavior.
+  private const float _spiralBearingThreshold = 10f;
```
[the remark this answers is not in the document]
Comment (tryuan99) [reply Assets/Scripts/Controller/PnController.cs]: I think eventually all constants should be configurable (but I think this can be addressed in a future PR).  Ideally, all parameters are in the configs that are loaded when starting the simulatio.

File: Assets/Scripts/Controller/IController.cs (hunk 5 above)
Comment (tryuan99) [reply Assets/Scripts/Controller/IController.cs]: Done in #116

**Review by daniellovell — APPROVED** [review]
